### PR TITLE
Show geomean pageload over all suites

### PR DESCRIPTION
**Pull request by klahnakoski**
This is currently in the proof-of-concept phase.  The important part is the `perfherder.js` file, which will setup a promise for each signature, and them use them as a cache.

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/compare/errors...packing?expand=1#diff-d92da353f761cd6870655190f89198a0

### CODE REVIEW

**Comment by klahnakoski** [comment]
This https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/357 to simplify the url parameter parsing.

**Comment by klahnakoski** [comment]
@armenzg the `const p = (asynch ()=>{...})();` pattern is a Promise, where `...` is run just once.  I am not sure if it can be made simpler.  Then it can be used as `await p;` (no parenthesis).

**Comment by klahnakoski** [comment]
@armenzg  

Do not review the code. Rather, I ask that you figure out how you would like this code organized so it can included, but somehow placed so you know it has scope limited to the `src/views/TP6mAggregate.jsx` that it supports.

In general, this code will serve as numeric computation library 

* **GMTDate** - solves the problem of defining a worldwide time reference, and performing operations on the that field of numbers.  Maybe there is a library for this, but the Javascript  ecosystem is still primitive.
* **Duration** - Works well with GMTDate.  This (and GMTDate) was copied, and converted to ES6. I am sure this can be replaced with a third party library, but there is a devil in the details: Ensuring math on  months is handled properly might be hard to find. 
* **Matrix** - very basic multi-dimensional array implementation.  It can be better implemented as a single array with algebra for [strides](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.strides.html), but it is easier to debug an actual array for now.
* **Cube** - A collection of matrices with named dimension access. The hard part of using Numpy ( or any matrix algebra) is not confusing the logical dimension and actual dimension.  A cube is a little like a Pandas dataframe, but simpler. 
* **Cube.window()** - Cube exists to support [window functions](https://mode.com/sql-tutorial/sql-window-functions/) over the muti-dimensional cubes. This is just a fancy `map()` function, which requires more parameters to orient itself along the dimension(s) it will be iterating.
* **Edge** - Used to describe an edge of the cube; there is one edge per dimension.
* **Domains** - each Edge has a domain that describes the range of values allowed
* **expressions.js** - used to compose expressions and turn them into callable functions.  It can not be seen now, but in the future, business rule (expressions) will be in configuration files and further manipulated by code to generate the proper requests to the backend datastores.
* **frum().edges()** - the path to multi-dimensional mode: Different than `groupBy` because it assumes the result is Cartesian product of the properties grouped by.

Some other notes:
* **new Array(length) is broken** - it seems ES6 has kept some *gothchas* from older Javascript: it is unsafe to make an Array and let it loose in the code; [it acts different from other arrays](https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/358/files#diff-acfab38101bb5fa5b4d8ffb2f6218578)
* **Logs** - until we decide on a logging library, I will use this one.

File: docs/Vendor use cases.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+## Definitions
+
+* **Project** - The collection of source code for a product
+* **Library** - Collection of functionality intended for use in projects
+
+Separate projects have own version of source
+
+* Bug fix library, no API change
+* Enhance library, API enhancement
+* Change library, API changed
```
Comment (armenzg) [inline docs/Vendor use cases.md]: In case you want to re-use terminology:
https://semver.org/

**Comment by klahnakoski** [comment]
The most important part of this PR is the `TP6mAggregate.jsx` file, which pulls the data and massages it to show the charts.
 
https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/358/files#diff-b081239380129dabeafba97c91342c0d

Showing under Raptor (TP6m):

https://deploy-preview-358--firefox-health-dashboard.netlify.com/android

**Comment by karlht** [comment]
Kyle, will you mark me as a reviewer for this PR, please?

File: src/vendor/jx/Edge.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable linebreak-style */
+/* eslint-disable no-underscore-dangle */
```
Comment (karlht) [inline src/vendor/jx/Edge.js]: (here and in other files) I think that every time you use an `eslint-disable` directive, some documentation is needed saying why.  If this code is ever to be maintained by anyone other than yourself, it should adhere to the Principle of Least Surprise.  For example, what is it that `eslint` doesn't like about your linebreak-style, and why is it important?

File: src/vendor/utils.js [hunk 3]
```diff
@@ -1,9 +1,22 @@
 /* global window */
 /* eslint-disable no-restricted-syntax */
 
+const { isArray } = Array;
+const zero = () => 0;
+
+function array(length = 0) {
```
Comment (karlht) [inline src/vendor/utils.js]: Narrative explanation of how `new Array()` is broken would be helpful in a comment here.

**Comment by klahnakoski** [comment]
@karlht 

The current name is `chainFrom` in an attempt to to better describe what it "does":  Actually, it does nothing but wrap an `Array` (or an iterable) to enable some useful actions.

My first attempt was to use Lodash. Lodash calls this method `chain()`.  I was told that Lodash idioms should not be used (https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/208#pullrequestreview-178774387)  which took me by surprise, because the alternative is to manually inline what Lodash does, or make my own Lodash-like set of functions. (https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/208#pullrequestreview-178774387).  

So the second attempt avoided `chain()`, but still used `_`; which is also not allowed  (https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/233#discussion_r241965123). I removed more Lodash and replaced it with `frum()`, which is a poor spelling of `from` which comes from SQL. I chose composition using a private `Wrapper()` class rather than inheriting from `Array` and making that class public. (https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/233/commits/1baed8b39d32d204a13e38898fc3e73fcf799798).

On the third attempt, the `frum()` function was unfamiliar to the other reviewers, along with other decisions. (https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/284) has more explanations. 

The fourth attempt, `frum()` required more explanation, plus a a historical review of the decisions (https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/291).  You can see Sarah first time suggesting `chainFrom`.  The fourth attempt, or soon after, was successful, despite we do not have agreement on the name of `frum()`, nor agreement on whether the whole class should exist.

Then while other work proceeded, I proposed `vector()`: https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/340 , which I thought was an amazing name, but turned out not to communicate well. I proposed `fluent`: Which has the benefit of being Googleable, and hints at why it would be useful. https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/359

Which brings us to this PR, using the Sarah's suggestion of `chainFrom()`.

What is `chainFrom()`?  It is a function that let you enter "chained vector operation mode".

File: src/config/mobileG5.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+const platforms = [
+  {platform: 'android-hw-g5-7-0-arm7-api-16'},
+  {platform: 'android-hw-g5-7-0-arm7-api-16-pgo'}
+];
+
+const fennec64 =
+  {
+    "header": ["geomean", "loadtime", "loadtime stdev", "fcp", "fnbpaint", "ttfi", "ttfi stdev", "dcf", "suite", "url"],
+
+
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/config/mobileG5.js]: There is a link at the top of the file

File: src/vendor/utils.js (hunk 3 above)
Comment (klahnakoski) [reply src/vendor/utils.js]: Each source file has a test of the same name. Here is the reason `array()` should be used instead of `new Array()`:

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/056032f7f8385a5f02246e8263f81d02414e0686/test/vendor/utils.test.js#L63

If you mean for others to better understand what this code does: Yes, I will add a comment.

File: src/vendor/jx/Edge.js (hunk 2 above)
Comment (klahnakoski) [reply src/vendor/jx/Edge.js]: Yes, that is an embarrassment you get to see with this WIP: I work in windows, and my IDE refuses to admit there are `\r` in my source files, even though there are. Git eventually scrubs them out, and I can remove this.

Even if this was permanent disabling: I would assume the file might not conform to `linebreak-style`, and I would assume it is unimportant because all linting is unimportant: Someone, somewhere wanted to disable linting rather than add more code to satisfy a linter.

I am still new at working for an aggressive linter.  So far, my experience is it provides copious amounts of low quality feedback, which pollutes the console output during debugging and interfering with finding legitimate errors. For me, each `eslint-disable` line is a signal that someone escaped oppression. :)

Seriously, I continue to look for a solution.

**Comment by klahnakoski** [comment]
@karlht 


On your three main points:

> Not enough narrative comments. The code shows "how," but it's difficult to see "why" in several places. 

Please add some "why?"s in the PR so I can get a sense of what answers you are looking for. I imagine all new relationships will require a number of interactions to establish rapport. Plus it will document this PR further. 

> the commit messages don't tell me anything that the diffs don't. Understanding your rationale will help me understand the code.

Armen has told me the recommended strategy is to squash the commit to effectively erase the developmental history of the feature so you need not read those commit messages. 

I believe squashing (or rebasing) is unwise. Merging is better because the chain-of-commits (the branch) is a recording of actual development: I use it to help evaluate student effort, and to recover pieces of my own attempts on my journey to a final solution. This means individual commits have little value.  Rather the value comes from the short story they tell to humans, and to inform git about how merge conflicts are to be resolved.

Only merge points matter; they have the complete diff and they come with a PR to describe the code change in more detail, which also contains review questions, like "why". 


> Do you intend to be sole maintainer of the new code going forward? If this code is going to have a significant lifetime, you'll want to think about succession plans and long-term maintenance.

I imagine library code has a few rough stages of maturity

1. **Utility** - The `vendor` files are currently utility functions in the Health dashboard. They will remain there for a while as they expand and optimize to the use cases found. Utility functions may have few tests because the application level tests provide codecoverage.
2. **Tool** - They apply for candidacy to the frontend-infra common toolbox: If accepted, they join a number of other common tools as code. The code is still actively developed by the larger team; more use cases are considered; and test suites are expanded. The tools in the toolbox are distributed as a single package.   
3. **Package** - Once reasonably mature, and architecture stops changing, the tool separates out to its own repo and package.

I hope that the files in the `src/vendor` directory are accepted as tools in the frontend-infra toolbox.  Until that happens, if ever, they will remain as utility functions in the Health dashboard; living only as long as there are dashboards that use them.

File: src/vendor/perf-goggles.js [hunk 5]
```diff
@@ -115,7 +114,7 @@ const transformOptionCollectionHash = optionCollectionHash => {
 };
 
 const treeherderOptions = async () => {
-  const optionCollectionHash = await queryAllTreeherderOptions();
+  const optionCollectionHash = await queryAllTreeherderOptions;
```
Comment (karlht) [inline src/vendor/perf-goggles.js]: (why) What's the semantic difference here?  Callable vs. not?

File: src/components/ChartJsWrapper/index.jsx [hunk 6]
```diff
@@ -48,24 +49,26 @@ const ChartJsWrapper = ({
       );
     }
 
-    let error = null;
-
-    data.datasets.forEach(dataset => {
-      const latestDataDate = new Date(Math.max(dataset.data.map(({ x }) => x)));
+    const allOldData = data.datasets.every(dataset => {
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/components/ChartJsWrapper/index.jsx]: yes, on line 65.

File: src/components/ChartJsWrapper/index.jsx [hunk 7]
```diff
@@ -117,8 +120,8 @@ ChartJsWrapper.propTypes = {
             x: PropTypes.oneOfType([
               PropTypes.string,
               PropTypes.instanceOf(Date),
-            ]).isRequired,
-            y: PropTypes.number.isRequired,
+            ]),
+            y: PropTypes.number,
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/components/ChartJsWrapper/index.jsx]: Some days do not have data, null data.  We want to say this day does not have data, which is different from this days does not exist; the rules for showing lines on the chart will be different.

File: src/vendor/logs.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
+/* eslint-disable max-len */
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/vendor/logs.js]: This `max-len` is for the whole file, specifically, line 8 and onward.  The comments exist to document the regex patterns: Removing the tail of the comments would loose information.  Allowing linewrap would be ugly.

File: src/vendor/math.js [hunk 9]
```diff
@@ -128,6 +128,16 @@ function average(values) {
   return total / count;
 }
 
+function exp(value) {
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/vendor/math.js]: I am not sure I get your point, but I am certain it is interesting.  The problem in this case is detailed in the `broken js: exp` test:

```
    // WE EXPECT THIS CALCULATION TO FAIL, OR RETURN NOTHING
    // WE DO NOT EXPECT A LEGITIMATE VALUE
    expect(Math.exp(null)).toBe(1);
```

There is missing data and I do not want missing data to give me results that look legitimate.

File: src/vendor/perf-goggles.js (hunk 5 above)
Comment (klahnakoski) [reply src/vendor/perf-goggles.js]: I have changed the definition of `queryAllTreeherderOptions`.  Before, it was an asynch function that you call to get Treeherder information. Now, it is a Promise to get Treeherder information.  The latter would make a real network call to get the data every time you called it.  The latter makes the call once (at creation time) and caches the result in itself for other promise chains to use. 

```
const promise = (asynch()=>{ /*do some network call here*/  })(); //only run once 
const result = await promise;  // use it anytime, and as many times as you like

File: src/vendor/math.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (karlht) [reply src/vendor/math.js]: Oh, for heaven's sake.  I didn't know that `Math.exp(null) === 1` was true.  That's just ... evil.

I was expecting `Math.exp(null)` to be `null`.  My mistake, carry on.  We'll save the spirited discussion for JavaScript type-coercion edge cases.

File: src/vendor/jx/expressions.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
+/* eslint-disable no-use-before-define */
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/vendor/jx/expressions.js]: Mutually recursive functions:  Specifically, the poorly named  `jx` function uses the `expressions` object and vice-versa.  I find the code is more elegant if each is defined once, as a declaration, rather than using procedural composition.  

Since `no-use-before-define` raises a question, I will add more code so I can remove the disable.

File: src/vendor/utils.js [hunk 11]
```diff
@@ -71,41 +113,59 @@ function isInteger(n) {
...
-  if (missing(val) || isArray(val)) {
-    return false;
-  }
-
-  return typeof val === 'function' || typeof val === 'object';
-}
+const OBJECT_CONSTRUCTOR = {}.constructor;
 
+/*
+ * objects that are best serialized as JSON objects
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/vendor/utils.js]: I am at a loss to describe it more; between the English saying it is a JSON object, and the code saying that data is any object that is a direct instance of `Object`, and the matching test suite with a number of examples.

```
  it('isData', () => {
    expect(isData({})).toBe(true);
    expect(isData([])).toBe(false);
    expect(isData(chainFrom([]))).toBe(false);
    expect(isData(() => 0)).toBe(false);
    expect(isData(GMTDate.now())).toBe(false);
    expect(isData(new Duration())).toBe(false);
    expect(isData('')).toBe(false);
    expect(isData('test')).toBe(false);
    expect(isData(undefined)).toBe(false);
    expect(isData(null)).toBe(false);
  });
```
I will make a `Data` class; even if not used, it will serve as a conceptual placeholder.

[the remark this answers is not in the document]
Comment (karlht) [reply src/vendor/utils.js]: A class might make a good illustration; a couple of extra cases in the test suite containing numeric and string literals, as well as non-empty lists and tuples/objects, would probably be sufficient.  The current test cases look a bit like "this is a list of the things that are not Data; everything else is."  If that's the intent, best to make it explicit.

File: src/vendor/perf-goggles.js (hunk 5 above)
Comment (karlht) [reply src/vendor/perf-goggles.js]: Ah!  Today I learned that Promises act as memoized procedures as well as bits of async code.  That makes things much clearer; thank you.

File: src/vendor/jx/expressions.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (karlht) [reply src/vendor/jx/expressions.js]: I think an explanatory comment indicating that `jx` and `expressions` are mutually recursive is probably sufficient here; I'll admit that I missed that fact on my first pass.  (This is one of those cases where I think I agree with you rather than with the linter, but I do think the explanatory comment would help anyone reading the code.)

File: src/components/ChartJsWrapper/index.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (karlht) [reply src/components/ChartJsWrapper/index.jsx]: This is precisely the sort of explanation I would love to see in a commit message.  Makes sense.

File: src/components/ChartJsWrapper/index.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (karlht) [reply src/components/ChartJsWrapper/index.jsx]: Man, my eyes are getting too old for two-space indents.  I knew it was there (or the code wouldn't have compiled) but ... yeah, reading multiply-indented things is definitely harder than it used to be.  Thanks.

[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/components/ChartJsWrapper/index.jsx]: Yes!  4-space would be much nicer!

File: src/vendor/Data.js [hunk 12]
```diff
@@ -110,9 +105,9 @@ Data.get = (obj, path) => {
       } else if (isInteger(step)) {
         output = output[step];
       } else if (isArray(output)) {
-        output = output.map(o => (isObject(o) ? o[step] : null));
+        output = output.map(o => (isData(o) ? o[step] : null));
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/vendor/Data.js]: More detail near definition of `isData`

**Comment by karlht** [comment]
In the spirit of SQL, I wonder if `selectFrom()` might get the point across.  Certainly in the `selectFrom(A).map(B).filter(C)` case, it has a superficial resemblance to SQL.

This may also be a case where an example is worth a thousand comments.  There's definitely enough code here to justify a README with workable examples.

File: src/vendor/utils.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/vendor/utils.js]: Done. I moved `isData` to `Data.js`, which has some other data manipulation functions, along with a new `Data` class:

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/358/files#diff-795f3e2d7bcca65d54e7515cb23c2163R25

I just learned that [`{[name]: value}` is legitimate syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Computed_object_property_names_and_destructuring)

**Comment by klahnakoski** [comment]
@armenzg are you able to see why the security thing fails?

image

**Comment by klahnakoski** [comment]
@davehunt  Please review. 

I am not asking for a code review. Rather, please look at the preview: https://deploy-preview-358--firefox-health-dashboard.netlify.com/

**Comment by klahnakoski** [comment]
@karlht   Another review please.

File: src/components/ChartJsWrapper/index.jsx [hunk 13]
```diff
@@ -120,15 +136,15 @@ ChartJsWrapper.propTypes = {
...
           })
         ),
         label: PropTypes.string.isRequired,
       })
     ).isRequired,
   }),
-  title: PropTypes.string,
+  title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/components/ChartJsWrapper/index.jsx]: The `title` can be some HTML that includes links to Perfherder data.  It is used here:

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/6a9e40fb2902e877395bdaa82f9052864274cbd2/src/containers/PerfherderGraphContainer/index.jsx#L230

File: src/quantum/Quantum.jsx [hunk 14]
```diff
@@ -149,27 +149,33 @@ export default class QuantumIndex extends React.Component {
...
-                    and: [seriesConfig, { eq: { test: 'loadtime' } }],
-                  },
+                .map(s => ({
+                  label: s.browser,
+                  seriesConfig: s.seriesConfig,
                 }))}
             />
           ))
-          .enumerate()
-          .limit(4),
```
[the remark this answers is not in the document]
Comment (klahnakoski) [reply src/quantum/Quantum.jsx]: In the old version, there was ?10? sites,  in popular-first order.  Since this is a summary page, the `limit(4)` was showing just the top 4 site timings. 

With all the sites added, keeping them in popular-first order was too messy: I put them in alphabetical order, and explicitly listed the 4 sites I want to show on this summary page

File: src/quantum/Quantum.jsx [hunk 15]
```diff
@@ -149,27 +149,33 @@ export default class QuantumIndex extends React.Component {
...
+        rows: selectFrom(TP6_COMBOS)
+          .where({
+            bits,
+            test: 'warm-loadtime',
+            site: [
+              'Tp6: Facebook',
+              'Tp6: Amazon',
+              'Tp6: YouTube',
+              'Tp6: Google',
+            ],
```
Comment (klahnakoski) [inline src/quantum/Quantum.jsx]: Just above are the four sites
